File: sem3/map/Exam/GUI_ToyLanguageInterpreter/src/Interpreter/Models/Expressions/ArithmeticExpression.java
package Interpreter.Models.Expressions;

import Interpreter.Models.ExceptionFactory;
import Interpreter.Models.Expression;
import Interpreter.Models.ProgramState.ProgramState;

public class ArithmeticExpression extends Expression {
    private Expression rightSide;
    private String operator;
    private Expression leftSide;

    public ArithmeticExpression(Expression leftSide, String operator, Expression rightSide) {
        this.leftSide = leftSide;
        this.operator = operator;
        this.rightSide = rightSide;
    }

    public int eval(ProgramState state) throws Exception {

        int rightSideValue = rightSide.eval(state);
        int leftSideValue = leftSide.eval(state);

        if (operator.equals("+")) {
            return leftSideValue + rightSideValue;
        }

        if (operator.equals("-")) {
            return leftSideValue - rightSideValue;
        }

        if (operator.equals("/")) {
            if (rightSideValue == 0) {
                throw ExceptionFactory.create("expression", "division by 0!");
            }
            return leftSideValue / rightSideValue;
        }

        if (operator.equals("*")) {
            return leftSideValue * rightSideValue;
        }

        if (operator.equals("%")) {
            if (rightSideValue == 0) {
                throw ExceptionFactory.create("expression", "division by 0!");
            }
            return leftSideValue % rightSideValue;
        }

        return 0;
    }

    @Override
    public String toString() {
        return leftSide.toString() + " " + this.operator + " " + rightSide.toString();
    }
}

File: sem3/map/Lab/toylanguage_C#/ToyLanguage/ToyLanguage/Models/Expressions/ArithmeticExpression.cs
using System;
using ToyLanguage.Models.Interfaces;
using ToyLanguage.Models.States;

namespace ToyLanguage.Models.Expressions
{
    public class ArithmeticExpression : IExpression
    {
        private readonly IExpression _leftSide;
        private readonly string _operator;
        private readonly IExpression _rightSide;

        public ArithmeticExpression(IExpression leftSide, string op, IExpression rightSide)
        {
            _leftSide = leftSide;
            _operator = op;
            _rightSide = rightSide;
        }
        
        public int Eval(ProgramState state)
        {
            var leftSideValue = _leftSide.Eval(state);
            var rightSideValue = _rightSide.Eval(state);

            if (_operator == "+")
            {
                return leftSideValue + rightSideValue;
            } 
            if (_operator == "-")
            {
                return leftSideValue - rightSideValue;
            }
            if (_operator == "*")
            {
                return leftSideValue * rightSideValue;
            }
            if (_operator == "/")
            {
                if (rightSideValue == 0)
                {
                    throw new Exception("Division by 0!");
                }
                return leftSideValue / rightSideValue;
            }

            return 0;
        }

        public override string ToString()
        {
            return $"{_leftSide} {_operator} {_rightSide}";
        }
    }
}